### PR TITLE
Don't expose `exception` error attribute by default

### LINKED ISSUE #7872: Consider not exposing `exception` error attribute by default
By default, `ErrorAttributes` will expose `exception` attribute which can be considered [information leakage](https://www.owasp.org/index.php/Top_10_2007-Information_Leakage_and_Improper_Error_Handling) similarly to #4730. IMO the `exception` attribute shouldn't be included by default but rather only using a mechanism similar to one used to include `trace` attribute.

To clarify using a sample:

```java
@RestController
class AppController {

        @GetMapping("/")
        String home() {
                throw new RuntimeException("Sample error message")
        }

}
```

Running this with Spring Boot CLI and hitting the `home` handler will yield:

```
HTTP/1.1 500 
Connection: close
Content-Type: application/json;charset=UTF-8
Date: Wed, 04 Jan 2017 16:34:22 GMT
Transfer-Encoding: chunked

{
    "error": "Internal Server Error", 
    "exception": "java.lang.RuntimeException", 
    "message": "Sample error message", 
    "path": "/", 
    "status": 500, 
    "timestamp": 1483547662254
}
```

### PR DESCRIPTION
**Pull request by vpavic**
This resolves #7872 by introducing `server.error.include-exception` configuration property which defaults to `false`.

I've also changed `DefaultErrorAttributes` to use `RequestDispatcher` constants instead of `String` literals for standard attribute names.

### CODE REVIEW

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java [hunk 1]
```diff
@@ -64,6 +66,12 @@
 	private static final String ERROR_ATTRIBUTE = DefaultErrorAttributes.class.getName()
 			+ ".ERROR";
 
+	private boolean includeException;
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: Shouldn't this be a constructor instead?

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java [hunk 2]
```diff
@@ -94,7 +102,7 @@ private void storeErrorAttributes(HttpServletRequest request, Exception ex) {
 	private void addStatus(Map<String, Object> errorAttributes,
 			RequestAttributes requestAttributes) {
 		Integer status = getAttribute(requestAttributes,
-				"javax.servlet.error.status_code");
+				RequestDispatcher.ERROR_STATUS_CODE);
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: That looks like an unrelated change to me.

File: spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributesTests.java [hunk 3]
```diff
@@ -180,9 +177,21 @@ private void testBindingResult(BindingResult bindingResult, Exception ex) {
 	}
 
 	@Test
+	public void exception() throws Exception {
```
Comment (snicoll) [inline spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributesTests.java]: Perhaps that method name could be a bit more explicit?

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java (hunk 1 above)
Comment (vpavic) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: Yes, I'd also prefer that. Just went with more passive approach initially, will update this.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java (hunk 2 above)
Comment (vpavic) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: Yes, that's what I was referring to in the opening comment:

> I've also changed DefaultErrorAttributes to use RequestDispatcher constants instead of String literals for standard attribute names.

Should I open a separate PR to address this?

File: spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributesTests.java (hunk 3 above)
Comment (vpavic) [inline spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributesTests.java]: Agreed.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java (hunk 2 above)
Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: We have a few places where we've avoiding referencing `RequestDispatcher` directly to help with Servlet 2.5 compatibility. My preference would be to take care of all of them in a single, separate change.

Comment (vpavic) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: OK - at which branch should I target that change?

Comment (wilkinsona) [inline spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/servlet/error/DefaultErrorAttributes.java]: Against `master`, please. But only if you're feeling particularly generous; it's not the most exciting of changes :)

**Comment by vpavic** [comment]
I've updated the PR to address the comments.

**Comment by snicoll** [comment]
This has failed the build (
`BasicErrorControllerIntegrationTests`) and I don't understand why the CI build worked... Looking into it.
